op-node-derivers.md

```diff
@@ -0,0 +1,554 @@
+# `op-node-derivers`
+
+#### Metadata
+
+Authors: @protolambda.
+Created: May 1, 2024.
+
+# Purpose
+
+<!-- *This section is also sometimes called “Motivations” or “Goals”.*
+ *It is fine to remove this section from the final document,
+  but understanding the purpose of the doc when writing is very helpful.* -->
+
+The core component of native cross-L2 interop is
+[message dependency-validation](https://github.com/ethereum-optimism/specs/blob/main/specs/interop/messaging.md).
+This is a new core chain-validation process, adjacent to derivation,
+that the rollup node implementation will have to support.
+
+The purpose of this design doc is to support the integration of new rollup-node processes like this,
+by improving the abstractions, encapsulating complexity, and decoupling chain-state from derivation-state better.
+
+# Summary
+
+<!-- *Most (if not all) documents should have a summary.
+While the length will likely be proportional to the length of the full document,
+the summary should be as succinct as possible.* -->
+
+To support "cross-unsafe" block validation of Interop:
+- Slim down and simplify the derivation, by decoupling the `EngineQueue`.
+- Encapsulate the driver sub-processes, by splitting the state-loop into smaller composable "derivers".
+
+
+# Problem Statement + Context
+
+<!-- *Describe the specific problem that the document is seeking to address as well as information
+needed to understand the problem and design space.
+If more information is needed on the costs of the problem, this is a good place to that information.* -->
+
+The op-node splits the "when" (driver) and the "how" (derivation).
+This separation of scheduling and processing is key for fault-proving,
+where there is no "when", and where the "how" needs to be synchronous.
+
+The problem is that neither driver or derivation is extensible,
+and with every feature that is forced into it, the total becomes more fragile and complicated.
+
+## Tech-debt
+
+The "tech-debt" topic applies to the op-node, not directly to the rollup-node spec.
+However, this is an area that is under-specified,
+and as de-facto reference implementation the op-node should not be encumbered with tech-debt that makes it unreadable.
+
+Things like the unsafe-block processing and sequencing work were fitted into safe-block processing code-path:
+1) This overloads the responsibilities of the derivation-pipeline, in particular the "engine-queue".
+2) This overloads the responsibilities of the driver state-loop, monolithically and synchronously executing everything.
+
+This is a remnant of the first Bedrock iteration, which synchronously processed L1 deposit events, and nothing else.
+
+This worked reasonably well up to now, since it was mostly centered around a single new resource: the "unsafe" block.
+
+The OP-Stack is growing a lot: new types of resources,
```

Comment:
Unless we modify the execution client, I think there are three resources we have to take a lock over
- Unsafe Head
- Safe Head
- Finalized Head

Block building is a bit weird b/c we should serialize those calls with the head updates, it can track it's own validity (i.e. is the payload still the tip payload & building on the same chain).

Reply:
Interop is already adding one (cross-unsafe), and previous designs of plasma was adding one (subjective-safe). And I think of the L1 traversal as another form of block-label resource too. For this reason I don't want to enshrine unsafe/safe/finalized as the only ones that matter.